Meta: Use superbuild to separate host and target builds

ADKaster opened the pull request:
Convert to superbuild infrastructure

Tasks:
- [x] Cleanup. Plenty of messing around smashing the keyboard until build says done artifacts left
- [x] Figure out what to do with all the custom targets from top lvl cmake lists (for target buld)
- [x] Create Toolchain file for Clang cross compiler (and like, actually test it)
- [x] Update serenity.sh
- [x] Make sure ports are not broken? x86_64 seems to not work atoll.
- [x] Rebase on master, and re-organize commits

This is based off of #9281

Majority of #4594 (some extra cleanup tasks in there that will be addressed after this.

Fixes #9425 in a drive-by commit at the end.

bgianfo commented on CMakeLists.txt:
```diff
@@ -48,7 +48,7 @@ option(USE_CLANG_TOOLCHAIN "Build the kernel with the experimental Clang toolcha
 
 add_custom_target(run
-    COMMAND ${CMAKE_COMMAND} -E env "SERENITY_ARCH=${SERENITY_ARCH}" ${CMAKE_SOURCE_DIR}/Meta/run.sh
+    COMMAND "${CMAKE_COMMAND}" -E env "SERENITY_ARCH=${SERENITY_ARCH}" "${SerenityOS_SOURCE_DIR}/Meta/run.sh"
```
nitpick: Why is the casing for `SerenityOS_SOURCE_DIR` special, shouldn't it match the casing for every other cmake variable and just be `SERENITYOS_SOURCE_DIR`?

ADKaster replied:
It's following the <ProjectName>_SOURCE_DIR convention.
https://cmake.org/cmake/help/latest/variable/PROJECT-NAME_SOURCE_DIR.html#variable:%3CPROJECT-NAME%3E_SOURCE_DIR

This one has ProjectName in the case that it was set in the call to `project()`. 

Rationale there is there's an option to Move the "top level" `CMakeLists.txt` file for Serenity into Meta/Serenity or some such, and have the "SuperBuild" be the top level. If we go with that route, then this pattern will make things easier. I made this change on the previous PR that's open, here.

https://github.com/SerenityOS/serenity/pull/9281#discussion_r684844975

bgianfo commented on Kernel/CMakeLists.txt:
```diff
@@ -432,32 +423,12 @@ if (ENABLE_KERNEL_ADDRESS_SANITIZER)
-
-# HACK: This is a workaround for CLion to grok the kernel sources.
-#       It's needed because CLion doesn't understand the way we switch compilers mid-build.
-add_compile_definitions(__serenity__)
```
Was this lost? I can't find it..

ADKaster replied:
In theory it's not required anymore. The Build/ARCH/CMakeToolchain.txt file should make CLion pick up the correct compiler from the start. I don't have a CLion install to test that theory though :)

bgianfo commented on Meta/CMake/serenity_toolchain.cmake:
```diff
@@ -0,0 +1,30 @@
+    # Note: We must generate the toolchain before calling project, as project looks for the toolchain entries
+    set(SERENITY_TOOLCHAIN "GNU" CACHE STRING "Compiler toolchain to use for Serenity (GNU or Clang)")
+    set(SERENITY_BUILD_DIR "${CMAKE_CURRENT_BINARY_DIR}")
+    set(SERENITY_SOURCE_DIR "${CMAKE_SOURCE_DIR}")
```
It's `SerenityOS_SOURCE_DIR` elsewhere, but `SERENITY_SOURCE_DIR` here? don't think that was something you introduced, but maybe since you are cleaning this all up, this could get fixed too?

ADKaster replied:
I tried a few approaches for the toolchain, including the "normal" way which seems to be environment varaibles? But doing that required setting SERENITY_SOURCE_DIR and SERENITY_ARCH every single time you did a rebuild, which is not very nice.

The only place these two variables are used is when generating the Toolchain file, like so:
```
set(triple @SERENITY_ARCH@-pc-serenity)
set(TOOLCHAIN_PATH @SERENITY_SOURCE_DIR@/Toolchain/Local/@SERENITY_ARCH@/bin)
set(TOOLCHAIN_PREFIX ${TOOLCHAIN_PATH}/${triple}-)

# where to read from/write to
set(CMAKE_SYSROOT @SERENITY_BUILD_DIR@/Root)
```
Up for bike shedding these guys though.

To answer the actual question, SerenityOS_SOURCE_DIR is a variable that gets set by he `project(SerenityOS)` call, and this toolchain cmake helper must be include()'d *before* that so that we can set CMAKE_TOOLCHAIN_FILE to the newly autogenerated toolchain file in the Build directory. The project() call is the place in the CMakeLists.txt that CMake detects what compiler/OS/etc we're compiling for, and where it reads CMAKE_TOOLCHAIN_FILE.

bgianfo replied:
That makes sense once explained, thanks!

friendlyanon commented on Meta/CMake/Superbuild/CMakeLists.txt:
```diff
@@ -0,0 +1,79 @@
+# https://stackoverflow.com/a/48555098
+# https://cmake.org/pipermail/cmake/2018-January/067002.html
+# MUST be done before call to 'project'
+get_cmake_property(vars CACHE_VARIABLES)
```
This is what I meant with my comment in the CI workflow. Provide the superbuild with a narrow contract and just document its usage. Are really that many variables shared by both that this approach is needed?

ADKaster replied:
There are about 10 options that are supported by both. Essentially:

- Sanitizer enable flags
- compile time format string flag
- enable all debug flag/all debug facilities flag for CI to ensure we don't have stale code behind -DFOO_DEBUG=1 (which was a problem I'm the past, and the reason all_the_debug_macros.cmake exists in the first place)
- flags that control downloading and using external resources. Ex. Wasm spec tests, Unicode character database, Unicode calendar database, etc. Still need to update the branch fully for the addition of the calendar database last week.

My biggest concerns are:

1) Some devs, esp those on macOS but some on older LTS Linux distros as well, have to pass a CMAKE_CXX_COMPILER today because their default host c++ compiler is not new enough to compile Lagom. Mostly AK/Format.h to be honest, though there's some concepts sprinkled around everywhere :). I really don't want them setting it for the superbuild configure though, or it'll get passed to the target build and screw everything up. Maybe that's a problem I created for myself by moving the Toolchain generation to the serenity build? The less options someone has the specify on the command line, the better imo :). 

2) It's unclear what the best way to tell CMake ok use your previous settings but add these and do the thing again. For example, if I just want to enable UBSAN on target, but keep everything else the same, how do I do that only interacting with the target build directory? I don't want to have to re-specify the CMake install prefix or common downloaded Unicode data file directory just to add ENABLE_UNDEFINED_SANITIZER=ON, or to set -DENABLE_EXTRA_KERNEL_DEBUG to get -O0 -ggdb3 for the kernel.

ADKaster replied:
Oh, and 3.

I really don't want to have to choose between visibility of options, and maintenance of multiple copies of the same option. I guess that's what Documentation/AdvancedBuildSettings.md (name?) Is for... But normally I remind myself what the options are by checking CMakeLists.txt. would be a shame to have to do it by checking common_options.cmake, then serenity_options.cmake or some such. Maybe that's a better design in general though 🤷‍♂️

friendlyanon replied:
> Some devs [...] have to pass a CMAKE_CXX_COMPILER today because their default host c++ compiler is not new enough to compile Lagom

The more convenient way to achieve this is to use the [CC](https://cmake.org/cmake/help/latest/envvar/CC.html) and [CXX](https://cmake.org/cmake/help/latest/envvar/CXX.html) env variables, which applies to more tools than just CMake as well.  
I regularly pass these variables to different builds without a second thought, so I guess this is more of a mindset kind of thing. I'm pondering whether more documentation would help here, but I'm still on the side of a narrow contract here.

> The less options someone has the specify on the command line, the better imo

Debatable. It can quickly become a slippery slope and you end up with a lot of hardcoded details that are now impossible to override from the CLI. Care must be taken to not end up there.

> It's unclear what the best way to tell CMake ok use your previous settings but add these and do the thing again

That's what the CMake cache is for:

```sh
# Let's build shared libs with ccache
cmake -S . -B build -G Ninja -D BUILD_SHARED_LIBS=1 -D CMAKE_CXX_COMPILER_LAUNCHER=ccache
# Actually, I wanted static libs, still build with ccache
cmake build -D BUILD_SHARED_LIBS=0
# Nope, need shared, with ccache, also export compile commands
cmake build -D BUILD_SHARED_LIBS=1 -D CMAKE_EXPORT_COMPILE_COMMANDS=1
```

> I really don't want to have to choose between visibility of options, and maintenance of multiple copies of the same option.

You won't have visibility of options of child builds from the superbuild. You can `ccmake`/`cmake-gui` the individual child builds to insecpt their cache for the variables. Maybe I'm misundersting you here.

ADKaster replied:
I didn't know that you could modify the CMake cache like that with `cmake <build_dir> <cache var to change>`. So, armed with that knowledge, it makes a heck of a lot more sense to use the Toolchain file properly instead of trying my hacky approach. I'll revert those hacks and move the cmake toolchain file generation into the superbuild instead, like I did the first go-around.

friendlyanon replied:
Regarding the common options, you can extract them to a cmake script:

```cmake
if(NOT COMMAND serenity_variable)
  macro(serenity_variable)
    set(${ARGV})
  endmacro()
endif()

serenity_variable(FOO BAR CACHE STRING "Documentaiton, ho!")
# ...
```

and in the superbuild, you can define the `serenity_variable` command ahead of including this, so you can collect the variables and their value automatically:

```cmake
set(common_options "")
macro(serenity_variable name value)
  set(${ARGV})
  list(APPEND common_options "-D${name}:STRING=${value}")
endmacro()

include(.../common-options.cmake)
```

This way you can list expand the `common_options` variable to either `CMAKE_CACHE_ARGS` or `CMAKE_CACHE_DEFAULT_ARGS`.

ADKaster replied:
that sounds good, but in order to let the cache for the superbuild affect the variable, I had to do this, with "-D${name}:STRING:=${${name}}"
```
# Collect options for Lagom build
set(lagom_options "")
macro(serenity_option name value)
    set(${ARGV})
    list(APPEND lagom_options "-D${name}:STRING=${${name}}")
endmacro()
```
push incoming, without a rebase :)

friendlyanon replied:
Yes, you are correct. I was in a bit of a rush to put that together. No need for the `value` parameter in that case.

friendlyanon commented on Meta/CMake/Superbuild/CMakeLists.txt:
```diff
@@ -0,0 +1,79 @@
+    SOURCE_DIR "${SERENITY_SOURCE_DIR}/Meta/Lagom"
+    BINARY_DIR "${PROJECT_BINARY_DIR}/../lagom"
+    INSTALL_DIR "${PROJECT_BINARY_DIR}/../lagom-install"
+    CMAKE_GENERATOR Ninja
```
This is not necessary, the generator used to generate the superbuild is proxied by EP already.

Same below as well.

ADKaster replied:
I actually don't want to use the same generator. The build takes around five minutes with a cold ccache on my machine. If the super build uses `ninja` then invoking `cmake --build Build/superbuild` does not show any progress on each sub-build. That is, the build freezes on Building serenity.... For five minutes then dumps the build log for all 3800 files into the terminal. Instead of having the nice progress bar go across the screen. 

One solution I found for this was too use Makefiles for the superbuild, and keep using ninja for the sub-builds. If there's another one I'm all ears :). But I really don't want to have people come into discord like "Andrew I updated to your new build approach and it just freezes after installing the host tools" when it's actually doing the build, they're just not seeing it.

friendlyanon replied:
> If the super build uses `ninja` then invoking `cmake --build Build/superbuild` does not show any progress on each sub-build.

That's weird, if you see the [fuzz step here](https://github.com/friendlyanon/cmake-init-gif-engine/runs/3450834777?check_suite_focus=true), you can see the output of both child builds, but also my [initial comment](https://github.com/SerenityOS/serenity/issues/4594#issuecomment-855310084) suggesting to move to a superbuild structure contained the output of Ninja builds that weren't spit out all at once.

A 5 minute build (assuming you have a beefy computer) doesn't sound terrible for my poor laptop, so I will investigate this one.

ADKaster replied:
https://user-images.githubusercontent.com/8388494/131226776-f50c6751-ac16-4072-b33e-722b5ca9e974.mp4

Here's the start of a superbuild build, note that it freezes until each stage is done, and then ninja dumps the entire output from that stage to the screen.

ADKaster replied:
https://user-images.githubusercontent.com/8388494/131227172-fb1251b9-72eb-4507-851e-ce356365c45c.mp4

Compared to an equivalent build with `Makefiles` as the generator. Gotta love `ccache` though :D.

friendlyanon replied:
I found the cause and it seems to be a peculiarity of the Ninja generators only. https://github.com/friendlyanon/cmake-init-gif-engine/commit/f91c21ed3c1b74844cce534d4fd0ede0ba0225e8

`ccache` is disgusting indeed 👀

friendlyanon commented on Meta/CMake/Superbuild/CMakeLists.txt:
```diff
@@ -0,0 +1,79 @@
+    EXCLUDE_FROM_ALL YES
+    CMAKE_ARGS ${LAGOM_CL_ARGS}
+    CMAKE_CACHE_DEFAULT_ARGS
+    "-DCMAKE_INSTALL_PREFIX:PATH=<INSTALL_DIR>"
+    "-DUNICODE_CACHE_LOCATION:PATH=${SERENITY_BUILD_DIR}"
+    "-DBUILD_LAGOM:BOOL=${BUILD_LAGOM}"
```
[Docs](https://cmake.org/cmake/help/latest/module/ExternalProject.html):

> The arguments are expected to be in the form `-Dvar:STRING=value`, which are then transformed into CMake `set()` commands ...

```suggestion
    -DCMAKE_INSTALL_PREFIX:STRING=<INSTALL_DIR>
    "-DUNICODE_CACHE_LOCATION:STRING=${SERENITY_BUILD_DIR}"
    "-DBUILD_LAGOM:STRING=${BUILD_LAGOM}"
```

This pattern is used by EP only to ensure that list values are properly processed.

Same below as well.

friendlyanon commented on Meta/CMake/Superbuild/CMakeLists.txt:
```diff
@@ -0,0 +1,79 @@
+)
+set(SERENITY_ARCH "i686" CACHE STRING "Architecture to build for (i686 or x86_64)")
+set(SERENITY_TOOLCHAIN "GNU" CACHE STRING "Compliler toolchain to use for Serenity (GNU or Clang)")
+set(SERENITY_BUILD_DIR "${PROJECT_BINARY_DIR}/../${SERENITY_ARCH}")
```
Going out of the binary directory?

ADKaster replied:
This was to make it work exactly the same as the old build, and have the Serenity build in `Build/<arch>`. There's three things this avoids breaking:

1) Meta/serenity.sh, most people's driver script for all things build (except I still see Andreas manually invoking ninja install && ninja image && ninja run in videos sometimes). I have some wip changes on the last commit to update that guy

2) Everyone's IDE settings. Looking around at the documentation for both CLion and VsCode (which should cover 90% or more of active devs) they don't particularly support configuring with a superbuild? So devs will have to manually point their IDE at, for example, the compile_commands.json in the build directory for Serenity. If we don't move that directory, it's one less thing to change in the documentation and a bit less painful to get this in.

3) Ports/. Hoo boy, ports. There's two main kinda of ports, the ones that use CMake where I can just update the path to the Toolchain file, and the ones that are mean and use autotools or some such. Things like prefix path, system include paths, etc are all baked into Ports/.port_include.sh and each individual Ports/port-name/package.sh. the most important ports to not break would probably be gcc, for running a compiler on target, make, CMake, git (?), and all the SDL2-xxx packages that use our fork of SDL2. It'd be nice to not touch those at the same time as trying to split out host tools from the target build.

friendlyanon replied:
Understandable, would be worth sprinkling a FIXME around this part with a reference to this comment or copy-pasting this into an issue of its own and referencing it.

ADKaster replied:
I've added a comment on the tagged line that references this thread.

friendlyanon commented on Meta/CMake/Superbuild/CMakeLists.txt:
```diff
@@ -0,0 +1,79 @@
+    "-DCMAKE_PREFIX_PATH:PATH=${PROJECT_BINARY_DIR}/../lagom-install"
+    "-DUNICODE_CACHE_LOCATION:PATH=${SERENITY_BUILD_DIR}"
+    "-DSERENITY_ARCH:STRING=${SERENITY_ARCH}"
+    "-DSERENITY_TOOLCHAIN:STRING=${SERENITY_TOOLCHAIN}"
```
[cmake-toolchains(7)](https://cmake.org/cmake/help/latest/manual/cmake-toolchains.7.html) file should be used instead. Any reason why that is not so?

ADKaster replied:
I am autogenerating the Toolchain files so that we don't have to set environment variables all over the place just to use it. The previous iteration of the Toolchain file was just for Ports/ and required setting both $SERENITY_ARCH and $SERENITY_SOURCE_DIR in order to do anything, which is tedious for an incremental build.

I'm generating it in the Serenity CMakeLists instead of in the superbuild because I wasn't sure how proscriptive I wanted to be about using the superbuild, and I hadn't looked at how to change Meta/serenity.sh to play nicely with it yet.

Most of that comes from not knowing the best way to tool up the superbuild. For example, I looked at creating cmake-presets, but it didn't look like IDEs of today would pick that up and see the host and target sub builds properly :/. Having a pre-configure step of, run CMake in the super build directory in your terminal before telling CLion/VsCode to use these exact settings for configuring the target build seems... Less than ideal? I hope I'm just missing something.

friendlyanon replied:
I will have to figure out a way to get my hands dirty with SerenityOS directly, but I still haven't switched to Linux host OS for pragmatic purposes on my laptop and the build tooling isn't exactly cross-platform friendly :^)

Yeah, you can forget about editors and IDEs directly recognizing superbuilds, it's not a common enough pattern that any of them bother with it. Consuming already generated child builds however is no problem, since they're just regular binary directories. Presets are "too new" and the existing tooling for those in editors and IDEs is terribad (of the ones I tried).

friendlyanon commented on Meta/CMake/Superbuild/CMakeLists.txt:
```diff
@@ -0,0 +1,79 @@
+    # Host tools must be built and installed before the OS can be built
+    DEPENDS lagom-install
+    STEP_TARGETS install
+)
```
Putting this at then end.  
Please consider adding https://github.com/friendlyanon/cmake-init-gif-engine/commit/662ca8894d833c8c945c537c5d97d95dc6ffa4cf to this as well. See the commit message for reasoning.

ADKaster replied:
Seems reasonable. I think the Makefile generator is currently broken for reasons related to the LibC build (go figure :D). But there's no sense having known bugs impact that and future(?) Visual Studio generator support.

friendlyanon commented on Meta/CMake/lagom-install-config.cmake.in:
```diff
@@ -1 +1,3 @@
+@PACKAGE_INIT@
+
```
```suggestion
```

This is not needed. Explained in `Lagom/CMakeLists.txt`.

friendlyanon commented on Meta/CMake/serenity_toolchain.cmake:
```diff
@@ -0,0 +1,30 @@
+#
```
Same comment as in `Superbuild/CMakeLists.txt`: Any reason why a proper toolchain is not used instead?

friendlyanon commented on Meta/CMake/utils.cmake:
```diff
@@ -48,16 +48,16 @@ function(serenity_shared_lib target_name fs_name)
     serenity_generated_sources(${target_name})
 endfunction()
 
+
+
```
Unnecessary whitespace.

friendlyanon commented on Meta/CMake/utils.cmake:
```diff
@@ -48,16 +48,16 @@ function(serenity_shared_lib target_name fs_name)
-    if (USE_CLANG_TOOLCHAIN)
-        target_link_libraries(${target_name} clang_rt.builtins-${SERENITY_CLANG_ARCH})
-    endif()
+    target_link_libraries(${target_name} "$<$<CXX_COMPILER_ID:Clang>:clang_rt.builtins-${SERENITY_CLANG_ARCH}>")
```
This is known at configure time. Use an `if()`.

friendlyanon commented on Meta/Lagom/CMakeLists.txt:
```diff
@@ -8,7 +8,14 @@ project(
     LANGUAGES C CXX
 )
 
+if (CMAKE_COMPILER_IS_GNUCXX AND CMAKE_CXX_COMPILER_VERSION VERSION_LESS 10.2)
```
Same as in `Kernel/CMakeLists.txt`.

friendlyanon commented on Meta/Lagom/CMakeLists.txt:
```diff
@@ -105,59 +109,51 @@ include_directories(${CMAKE_BINARY_DIR})
-        DESTINATION "${Lagom_INSTALL_CMAKEDIR}"
-        COMPONENT Lagom_Development
-    )
+configure_package_config_file(
```
`configure_package_config_file` is more of a utility function for those projects that have a little more advanced config file with package components and other needs. As it stands now, the config file is just a proxy to the exported target set, so reaching for this function is unnecessary. It also depends on you providing the install path at configure time, but that can be changed at install time:

```sh
cmake .. -D CMAKE_INSTALL_PREFIX=foo # install to foo
cmake --build .
cmake --install . --prefix bar # actually no, install to bar
```

ADKaster replied:
Gotcha. I had added that @PACKAGE_INIT@ and configure_package_config_file when trying to debug why things weren't working and grasping for straws. Turns out the solution to why it wasn't finding the targets was that the host build was x86_64 and the target build was i686 and I needed to specify ARCH_INDEPENDENT on the package version file creator. So no qualms in removing it.

friendlyanon replied:
Technically you don't even need versioning either, since (seemingly) nothing else is versioned in SerenityOS. That's a good reminder. I don't see a scenario where `find_package(Lagom <version>)` would be useful.

ADKaster replied:
Eh, true, the policy of the project is very much Live-At-Head and Live-In-Tree. I'm inclined to leave it under the assumption that CMake will get upset if it's not present?

friendlyanon replied:
No, the version config is completely optional. The only time CMake ever looks for it is when a version is requested and when it's not found then CMake will simply consider that particular package unsitable and move on to the next one.

[Docs](https://cmake.org/cmake/help/latest/command/find_package.html#version-selection):

> If no such version file is available then the configuration file is assumed to not be compatible with any requested version.

friendlyanon commented on Toolchain/CMake/ClangToolchain.txt.in:
```diff
@@ -0,0 +1,40 @@
+set(CMAKE_RANLIB ${TOOLCHAIN_PATH}/llvm-ranlib)
+set(CMAKE_STRIP ${TOOLCHAIN_PATH}/llvm-strip)
+set(CMAKE_AR ${TOOLCHAIN_PATH}/llvm-ar)
+set(CMAKE_CXXFILT ${TOOLCHAIN_PATH}/llvm-cxxfilt)
```
https://github.com/SerenityOS/serenity/issues/9425

ADKaster replied:
ah, reserved words. Fun in every language. I've added a commit to fix this.

friendlyanon commented on Toolchain/CMake/ClangToolchain.txt.in:
```diff
@@ -0,0 +1,40 @@
+
+set(CMAKE_EXE_LINKER_FLAGS_INIT "-Wl,--hash-style=gnu,-z,relro,-z,now,-z,noexecstack,-z,max-page-size=0x1000")
+
+set(CMAKE_C_COMPILER_WORKS YES)
+set(CMAKE_CXX_COMPILER_WORKS YES)
```
Why is this needed? This shouldn't generally be messed with.

ADKaster replied:
We use the Toolchain file to build things required to link a real executable, such as startup files (Userland/Libraries/LibC/crt0_shared.cpp for example).  So from a clean build it won't work. I was thinking of replacing this with the flag that says "use a static lib to check if the compiler works".

friendlyanon replied:
These variables come from [here](https://github.com/Kitware/CMake/blob/master/Modules/CMakeTestCCompiler.cmake#L38) and [here](https://github.com/Kitware/CMake/blob/master/Modules/CMakeTestCXXCompiler.cmake#L38). Can the toolchain really not compile those simple test sources?

ADKaster replied:
```
CMake Error at /usr/share/cmake-3.17/Modules/CMakeTestCCompiler.cmake:60 (message):
  The C compiler

    "/home/andrew/serenity/Toolchain/Local/clang/i686/bin/clang"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: /home/andrew/serenity/Build/i686/CMakeFiles/CMakeTmp

    Run Build Command(s):/usr/bin/ninja cmTC_2f4eb && [1/2] Building C object CMakeFiles/cmTC_2f4eb.dir/testCCompiler.c.o
    [2/2] Linking C executable cmTC_2f4eb
    FAILED: cmTC_2f4eb
    : && /home/andrew/serenity/Toolchain/Local/clang/i686/bin/clang --target=i686-pc-serenity --sysroot=/home/andrew/serenity/Build/i686/Root   -Wl,--hash-style=gnu,-z,relro,-z,now,-z,noexecstack,-z,max-page-size=0x1000 CMakeFiles/cmTC_2f4eb.dir/testCCompiler.c.o  -o cmTC_2f4eb   && :
    ld.lld: error: cannot open crt0.o: No such file or directory
    ld.lld: error: cannot open crti.o: No such file or directory
    ld.lld: error: unable to find library -lc
    ld.lld: error: cannot open crtn.o: No such file or directory
    clang-12: error: linker command failed with exit code 1 (use -v to see invocation)
    ninja: build stopped: subcommand failed.
```
Nope it cannot. LibC comes from the build, crt0.o comes from the build, as does crti.o and crtn.o.

We provide crti.o and crtn.o from the asm sources [here](https://github.com/SerenityOS/serenity/tree/fa4eeca0ac3330ba6fee226b315ca06652a7ea11/Userland/Libraries/LibC/arch/i386), while crt0.o comes from [this cpp file](https://github.com/SerenityOS/serenity/blob/fa4eeca0ac3330ba6fee226b315ca06652a7ea11/Userland/Libraries/LibC/crt0.cpp)

So we really don't have the things necessary to run those try compile checks before building LibC. 

So, going to probably replace this with [CMAKE_TRY_COMPILE_TARGET_TYPE](https://cmake.org/cmake/help/latest/variable/CMAKE_TRY_COMPILE_TARGET_TYPE.html#cmake-try-compile-target-type) set to STATIC.

friendlyanon replied:
Commit that introduces CMAKE_TRY_COMPILE_TARGET_TYPE. https://github.com/Kitware/CMake/commit/7f1bd9fe6910f7633d98dec018cc01331a46b87e

> This will be useful for cross-compiling toolchains that cannot
> link a binary without custom flags or scripts.

I don't believe that's true for SerenityOS, this feels like it'd be a hack. These object files could be built ahead of time, before the OS child build tree is configured.  
If I'm reading your reply correctly, it sounds like LibC should be moved into its own build as well.

ADKaster replied:
I'm not sure I'm seeing the value in creating yet another toolchain file and build stage right now just for the C Library build. That build would definitely need that TRY_COMPILE_TARGET_TYPE setting, because it's building the C library.

I can leave a FIXME to take a look at that. There's definitely some things that LibC/CMakeLists.txt does that are questionable.. like creating explicit static and shared libraries, etc. We include separate object libs for like. `ubsan` and `ssp` for `__stack_chk_fail` and some ... interesting add_dependencies there that make it a bit of a ball of wax on its own.

friendlyanon replied:
I see. Aiming for the cleanest possible solution would be more effort than worth right now. Using the `STATIC_LIBRARY` option with a FIXME is a good compromise for the sake of incremental improvement.

ADKaster replied:
For some historical context, we used to actually build the C library before libstdc++ as part of the toolchain build, because it was observed that libstdc++ actually needed a working C library in order to compile/link. But, turns out that it only needed the headers after Ben looked into it some more.

https://github.com/SerenityOS/serenity/commit/5f724b6ca1aae3a5a8c7189069649e8a9347cca2#diff-27e8b38fa0c632f71ab59fc31a34847ece5cbbc670f74514d313217fb25637a3L244-R252

But supporting that pre-libstdc++ build caused quite a few headaches, as things like `std::initailizer_list`, or `__cxa_demangle` were not available. Since that time we've moved to CMake, so the build tool logic to pre-build LibC seems to have been lost to time (and like > 10k commits).

friendlyanon commented on Userland/DynamicLoader/CMakeLists.txt:
```diff
@@ -36,11 +36,9 @@ set_source_files_properties(../Libraries/LibC/stdio.cpp PROPERTIES COMPILE_FLAGS
-endif()
-
+target_link_libraries(Loader.so
+    "$<$<CXX_COMPILER_ID:Clang>:clang_rt.builtins-${SERENITY_CLANG_ARCH}>"
+    "$<$<CXX_COMPILER_ID:GNU>:gcc>"
```
Use `if()`.

friendlyanon commented on Meta/CMake/serenity_toolchain.cmake:
```diff
@@ -0,0 +1,30 @@
+    configure_file("Toolchain/CMake/${SERENITY_TOOLCHAIN}Toolchain.txt.in" CMakeToolchain.txt @ONLY)
+    set(CMAKE_TOOLCHAIN_FILE "${CMAKE_CURRENT_BINARY_DIR}/CMakeToolchain.txt" CACHE PATH "Toolchain file to use for cross-compilation")
+
+    # Note: This is the price we pay for not requiring the toolchain to be generated externally
+    #       If the toolchain is specified via command line, we get its local variables imported into our CMake variables
```
This note is kind of silly now with the superbuild structure. You have plenty of opportunities to generate the toolchain and pass is to the build that needs it.

friendlyanon reviewed (changes requested):
Considering that this is just a draft for now, I have not commented on everything I wanted to, because there are a few questions still.

ADKaster commented on Kernel/CMakeLists.txt:
```diff
@@ -349,40 +349,31 @@ set(SOURCES
-if (USE_CLANG_TOOLCHAIN)
-    add_compile_options(-Waddress-of-packed-member)
-endif()
+add_compile_options("$<$<CXX_COMPILER_ID:Clang>:-Waddress-of-packed-member>")
```
Oh, I thought that using the generator expression would be preferred to using an if(). Is that not the case? I can change all these back to ifs, but based based on CMAKE_CXX_COMPILER_ID instead of the custom USE_CLANG_TOOLCHAIN variable then

friendlyanon replied:
It's just not really necessary to use a generator expression when what you are querying is already known at configure time. Genexes can also become quite unwieldy quickly and `if()` is just more readable.

friendlyanon reviewed (changes requested):
Could you please avoid rebasing until the end of the PR? Makes tracking review comments difficult, since they are not anymore linked to the changes. You can move and fixup commits in interactive rebase when things are done.

ADKaster commented on Kernel/CMakeLists.txt:
```diff
@@ -360,44 +360,42 @@ else()
-if (USE_CLANG_TOOLCHAIN)
+if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+    # Apply any flags that are only available on >= GCC 11.1
+    if (CMAKE_COMPILER_IS_GNUCXX AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 11.1)
```
D'oh, copy pasted this one up here earlier in the lists file without removing the old deprecated variable.

friendlyanon replied:
And just as a justification for quoting the version number, here is a related bugfix from CMake itself. https://github.com/Kitware/CMake/commit/49a5dbcaff57da133aadd7a31abdd65515b426e6

ADKaster commented:
> Could you please avoid rebasing until the end of the PR? Makes tracking review comments difficult, since they are not anymore linked to the changes. You can move and fixup commits in interactive rebase when things are done.

Sure thing, the diffs b/w revisions on github isn't particularly good with rebases, that's for sure. I still have a lot of planned changes to this PR though, one of which is splitting the first commit into multiple atomic commits (adding the `lagom_tool` cmake function, making the host tool consuming functions use an alias in the Lagom:: namespace, etc). Part of the code submission policy is to "Squash your commits when making revisions after a patch review", so that part came as a bit of a reflex 😅 .

friendlyanon commented:
> Part of the code submission policy is to "Squash your commits when making revisions after a patch review"

Yes, but that can happen before merging and as you mentioned, GitHub's UI is not the greatest for tracking changes when there are rebases, this PR will most likely be iterated on a "bit" more, so I thought I would request rebases to be left to the end of things.

ADKaster commented on .github/workflows/cmake.yml:
```diff
@@ -134,22 +133,21 @@ jobs:
         # Build the entire project with all available debug options turned on, to prevent code rot.
         # However, it is unweildy and slow to run tests with them enabled, so we will build twice.
-        run: cmake .. -GNinja -DSERENITY_ARCH=${{ matrix.arch }} -DBUILD_LAGOM=ON -DENABLE_ALL_DEBUG_FACILITIES=ON -DENABLE_PCI_IDS_DOWNLOAD=OFF -DENABLE_USB_IDS_DOWNLOAD=OFF -DCMAKE_C_COMPILER=gcc-10 -DCMAKE_CXX_COMPILER=g++-10
+        run: cmake -S ./Meta/CMake/Superbuild -B ./Build/superbuild -GNinja -DSERENITY_ARCH=${{ matrix.arch }} -DBUILD_LAGOM=ON -DENABLE_ALL_DEBUG_FACILITIES=ON -DENABLE_PCI_IDS_DOWNLOAD=OFF -DENABLE_USB_IDS_DOWNLOAD=OFF -DCMAKE_C_COMPILER=gcc-10 -DCMAKE_CXX_COMPILER=g++-10
```
I think I'm missing a step here, because right after creating the superbuild directory, there isn't a CMakeCache.txt file in the child build binary directories.

friendlyanon replied:
Right, you caught me with my pants down. A build must happen first before those directories are populated. Now my initial suggestion just makes me look silly.

I guess this can be resolved and continued in the [discussion for Superbuild/CMakeLists.txt](https://github.com/SerenityOS/serenity/pull/9297#discussion_r697862017).

ADKaster commented:
Phew. Changes since last review: 

- Option strategy implemented
- Toolchain strategy implemented
- Comments about formatting, quotes, etc, addressed
- Docs written
- Meta/serenity.sh in a reasonable state for the new ~~world order~~ build strategy.

In order to pull this out of draft I'll absolutely have to rebase and squash and organize the commits though, there's conflicts :(

friendlyanon reviewed (changes requested):
> In order to pull this out of draft I'll absolutely have to rebase and squash and organize the commits though, there's conflicts :(

One smart way of doing this kind of fixes after the fact is rewinding to a commit with interactive rebase using the `edit` operation, adding your changes, stashing the changes, continuing the rebase, then repeating until your `edit` commits are all done.  
Once there, you can just pop changes one by one from the stash stack and commit them individually in the order you stashed them.  
When it comes to merging these fixes back, you just move the commits in interactive rebase where you originally stashed them and mark them for `fixup` or whatever else.

ADKaster commented on Meta/Lagom/CMakeLists.txt:
```diff
@@ -30,8 +23,26 @@ get_filename_component(
+    include(all_the_debug_macros)
+endif(ENABLE_ALL_THE_DEBUG_MACROS)
+
+option(BUILD_SHARED_LIBS "Build shared libraries instead of static libraries" ON)
```
I'm not sure I understand the rationale here. Why not change our build to default to shared libraries? I don't want the CMake default of static libraries from Lagom and to force us to pass BUILD_SHARED_LIBS=ON everywhere. Everywhere being in Meta/serenity.sh, the docs, and in the libjs-test262 repo Linus manages.

We only want static libs in oss-fuzz where changing to shared requires a pull request to our config file there (which isn't the biggest deal but still an extra step).

ADKaster replied:
I don't intend to fix this right now, as it affects two external projects. I created #9867 to track this CMake best practice improvement.

Ideally we could move all of the oss-fuzz build scripting into a `Meta/Lagom/oss-fuzz-build.sh` or some such to make that less of a pain. Probably will be part of a larger oss-fuzz refactor, b/c currently we don't use our generated unicode tables when fuzzing (#9466).

friendlyanon replied:
I see, if it makes things simpler in the grand scheme of things, then so be it.

Personally, I would like to see a list of built-ins that are set to a non-default value, but that's maybe for another PR, if ever.  
Except the ones in toolchains. Toolchain files are (or should be) pretty self explanatory.

ADKaster commented on Documentation/BuildInstructions.md:
````diff
@@ -7,7 +7,7 @@ Make sure you have all the dependencies installed:
 
 ```console
-sudo apt install build-essential cmake curl libmpfr-dev libmpc-dev libgmp-dev e2fsprogs ninja-build qemu-system-i386 qemu-utils ccache rsync genext2fs
+sudo apt install build-essential cmake curl libmpfr-dev libmpc-dev libgmp-dev e2fsprogs ninja-build qemu-system-i386 qemu-utils ccache rsync genext2fs unzip
````
In master I recently added a FIXME for using file(EXTRACT_ARCHIVE) when we move to CMake 3.18 for all the fun archive extractions we currently do during the configure step that explicitly call out `tar` and `unzip` (while also adding a find_program for those two tools).

Though reading the docs there maybe they should be changed to `"${CMAKE_COMMAND}" tar` instead?

ADKaster replied:
Opened #9866 for this

ADKaster commented on Documentation/AdvancedBuildInstructions.md:
```diff
@@ -53,18 +59,71 @@ Many parts of the SerenityOS codebase have debug functionality, mostly consistin
 
-For the changes to take effect, SerenityOS needs to be recompiled and the disk image needs to be rebuilt.
+Note that changing CMake cache variables such as PROCESS_DEBUG and ENABLE_UNDEFINED_SANITIZER can be done on
+an existing build without reconfiguring completely, and their effect will take effect the next time SerenityOS is recompiled.
```
It's not just news to me, Andreas was unaware of this CMake concept in his latest office hours :) (around 1:15:50 https://youtu.be/puTsBKAQzk4?t=4552).  pointed it out to him on discord and he said "aha, I'd wondered why CMake options appeared 'sticky'". 

So I definitely want to put some sort of note in that conveys that CMake build directories do indeed cache their options/variables and that you can reconfigure to change just the specified from the command line while keeping the rest intact.

friendlyanon replied:
I'm in no way against documenting this, just how it is right now feels overly elaborate.

Pointing out the fact that there is `ccmake`/`cmake-gui` for browsing the cache would also be worth it.  
Maybe a separate section dedicated to this would be better than a blurb at the bottom of this section.

ADKaster replied:
Ah that''s a good point. I've personally never invoked `ccmake` or `cmake-gui` for browsing the cache though, so I don't think I'm the best person to describe how that works in a detailed section :thinking: . It's possible that some of the manual work that has gone into getting Meta/Lagom/Tools/ConfigureComponents to look pretty in `whiptail` could be replaced by just using `ccmake`.

friendlyanon replied:
The TUI and GUI are pretty easy to use. You just need to invoke them on the binary directory:
```console
$ ccmake path/to/existing/binary/dir
```
or in the GUI's case, you have the usual GUI stuff to do that.

I don't know what whiptail is, but if you love ncurses TUIs then ccmake will be right up your alley, even if a bit basic.

ADKaster replied:
whiptail is a TUI dialog box creator, the way we're using it creates a KConfig-looking interface for trimming down the system components that get created.

https://user-images.githubusercontent.com/8388494/132270074-d3210628-b253-4518-ad61-cf4cbe70cd55.mp4

ConfigureComponents is basically just a way to set the CMake cache variables for `BUILD_COMPONENTNAME` within some organized groups like "REQUIRED" "OPTIONAL" and "RECOMMENDED".

anyway, I'll move the cache description to a separate section with a blurb about the builtin tools for manipulating it alongside using `cmake` directly. After that i'm gonna start rebasing and splitting commits up so we can get this guy out of draft status.

ADKaster commented on Meta/Lagom/CMakeLists.txt:
```diff
@@ -30,8 +23,26 @@ get_filename_component(
+
+option(BUILD_SHARED_LIBS "Build shared libraries instead of static libraries" ON)
+if (ENABLE_OSS_FUZZ)
+    set(BUILD_SHARED_LIBS OFF) # Don't use shared libraries on oss-fuzz, for ease of integration with their infrastructure
```
I changed it a forced cache variable, the proper solution can be implemented in #9867

ADKaster commented:
Phew, rebased and updated for some unexpected upstream changes (mostly the support for clang toolchain to Meta/serenity.sh).
Assuming the CI passes this'll be ready for prime time review. Probably gonna have to force push some CI fixes though, there's always CI fixes :(

friendlyanon commented on Meta/Lagom/CMakeLists.txt:
```diff
@@ -24,16 +24,43 @@ get_filename_component(
+
+if (ENABLE_ALL_THE_DEBUG_MACROS)
+    include(all_the_debug_macros)
+endif(ENABLE_ALL_THE_DEBUG_MACROS)
```
Oh no, let's not repeat the condition in the ending command.

ADKaster replied:
Oh, true. not sure how that syntax got like that. probably copy  pasted from the main cmake lists and fixed the style in the original. fixed in ea349c1e54a2d17a84219347401eef99cf050790

friendlyanon commented on Meta/Lagom/Tools/CodeGenerators/LibWeb/CMakeLists.txt:
```diff
@@ -1,11 +1,6 @@
-add_executable(Generate_CSS_ValueID_h Generate_CSS_ValueID_h.cpp)
-add_executable(Generate_CSS_ValueID_cpp Generate_CSS_ValueID_cpp.cpp)
-add_executable(WrapperGenerator WrapperGenerator.cpp)
+lagom_tool(Generate_CSS_PropertyID_h   SOURCES Generate_CSS_PropertyID_h.cpp)
```
Since this function is taking the `SOURCES` variable from the enclosing scopes, I would recommend putting a `set(SOURCES "")` at the beginning on this lists file, to avoid any chance of accidentally including unrelated sources.

ADKaster replied:
Good call, fixed in d6c4ea86018ee722aae5b0b4d34e81e28313b4c6

friendlyanon commented on Meta/Lagom/Tools/CodeGenerators/StateMachineGenerator/CMakeLists.txt:
```diff
@@ -2,5 +2,4 @@ set(SOURCES
 
-add_executable(StateMachineGenerator ${SOURCES})
-target_link_libraries(StateMachineGenerator LagomCore)
+lagom_tool(StateMachineGenerator ${SOURCES})
```
I don't think this was intentional. `SOURCES` is taken from the calling scope(s) already.

ADKaster replied:
Indeed it wasn't. incomplete transformation on this file. fixed in d6c4ea86018ee722aae5b0b4d34e81e28313b4c6

friendlyanon commented on Toolchain/CMake/GNUToolchain.txt.in:
```diff
@@ -0,0 +1,41 @@
+
+set(CMAKE_SYSTEM_NAME SerenityOS)
+set(CMAKE_SYSTEM_PROCESSOR "@SERENITY_ARCH@")
+set(CMAKE_CROSS_COMPILING 1)
```
This is UB. Please see the documentation of [CMAKE_CROSSCOMPILING](https://cmake.org/cmake/help/latest/variable/CMAKE_CROSSCOMPILING.html).

ADKaster replied:
removed in ea349c1e54a2d17a84219347401eef99cf050790

friendlyanon commented on Toolchain/CMake/ClangToolchain.txt.in:
```diff
@@ -0,0 +1,42 @@
+
+set(CMAKE_SYSTEM_NAME SerenityOS)
+set(CMAKE_SYSTEM_PROCESSOR "@SERENITY_ARCH@")
+set(CMAKE_CROSS_COMPILING 1)
```
This is UB. Please see the documentation of [CMAKE_CROSSCOMPILING](https://cmake.org/cmake/help/latest/variable/CMAKE_CROSSCOMPILING.html).

ADKaster replied:
removed in ea349c1e54a2d17a84219347401eef99cf050790 . think this is something I set trying to get the try_compile checks  we talked about previously "working" and forgot to remove.

friendlyanon reviewed (changes requested):
Just a few small things. After these fixes, I think we should be done.

friendlyanon commented:
One interesting test to run would be seing how this change would affect an existing binary directory. I'm not sure how prevalent keeping the binary directory around in case of bigger changes is, maybe you can ask people on Discord?  
Seeing how little experience most people have with build systems, I'm guessing blowing away the binary directory upon the slightest misbehavior is the most common method of "fixing" things.

ADKaster commented on Documentation/AdvancedBuildInstructions.md:
```diff
@@ -43,27 +49,94 @@ There are some optional features that can be enabled during compilation that are
+
+The main limitation of the SuperBuild is that any non-option CMake cache variables such as component configuration or debug flag settings must be done
+after a build has started. That is, the CMakeCache.txt for the Serenity and Lagom builds is not created until the SuperBuild build starts and reaches the
+proper stage for the build in question.
```
Updated the end of the docs here with a link to the cache section and an example of modifying each child build in c15292c33d2161ad9625e8dea1d280b0aeb0d773

ADKaster commented:
> One interesting test to run would be seeing how this change would affect an existing binary directory. 

It gets very upset, and hits my FATAL_ERROR about CMAKE_SYSTEM_NAME

```
[11/15] Performing configure step for 'serenity'
loading initial cache file /home/andrew/serenity/Build/superbuild-i686/serenity-prefix/tmp/serenity-cache-.cmake
CMake Error at CMakeLists.txt:15 (message):
  System name is not SerenityOS, this is unsupported.

  Please re-read the BuildInstructions documentation, and use the superbuild
  configuration
```

Seems people will need to blow away their build directories. Their old caches will be quite a bit messed up :) Which to be honest isn't that big a deal, every time someone changes AK/Platform.h or AK/StdLibExtras.h or AK/String.h or... we need like a 3k+ file rebuild. Which seems to happen about once a week.

friendlyanon commented on Meta/CMake/Superbuild/CMakeLists.txt:
```diff
@@ -0,0 +1,137 @@
+    BINARY_DIR "${PROJECT_BINARY_DIR}/../lagom"
+    INSTALL_DIR "${PROJECT_BINARY_DIR}/../lagom-install"
+    EXCLUDE_FROM_ALL YES
+    CMAKE_CACHE_DEFAULT_ARGS
```
Now that I'm looking, you are passing the cache options using the `DEFAULT` variant here, which means that after the first configure with the superbuild, the values will not be propagated, because these are just defaults. The `CMAKE_CACHE_ARGS` argument should be used, which will use the `FORCE` option when setting the variables. I think all of them can be passed via `CMAKE_CACHE_ARGS`, because they are either cache variables in the superbuild or are dependent on code in this lists file.

friendlyanon reviewed (approved):
Just one small change, but otherwise LGTM!

ADKaster commented:
Updates since last push:

- Update CLion docs, needed to tweak the args passed to the Profile there to get things playing nice.
- Remove CMAKE_C_COMPILER_TARGET, CMAKE_CXX_COMPILER_TARGET and CMAKE_ASM_COMPILER_TARGET from GNUToolchain.txt.in. For some reason CLion was doing some weird things with these flags and choked on trying to determine the compiler version for itself by.. passing it too many times? or something. Those variables are not needed for the GNU toolchain, so just don't set them. they are 100% required for the clang toolchain though.

alimpfard reviewed (approved):
Ell Gee Tee Em!
